File: Sources/RMBTQoSTestGroup.m
/*
 * Copyright 2017 appscape gmbh
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#import "RMBTQoSTestGroup.h"
#import "RMBTQoSDNSTest.h"
#import "RMBTQoSHTTPTest.h"
#import "RMBTQoSTracerouteTest.h"
#import "RMBTQoSWebTest.h"
#import "RMBTQoSUDPTest.h"
#import "RMBTQoSTCPTest.h"
#import "RMBTQoSNonTransparentProxyTest.h"
#import "RMBT-Swift.h"

typedef RMBTQoSTest* (^RMBTTestQoSInitializerBlock)(NSDictionary*);

@interface RMBTQoSTestGroup() {
    RMBTTestQoSInitializerBlock _initializer;
}
@end

@implementation RMBTQoSTestGroup

- (instancetype)initWithKey:(NSString*)key
       localizedDescription:(NSString*)description
                initializer:(RMBTTestQoSInitializerBlock)initializer {
    if (self = [super init]) {
        _key = key;
        _localizedDescription = description;
        _initializer = initializer;
    }
    return self;
}

-(RMBTQoSTest*)testWithParams:(NSDictionary*)params {
    RMBTQoSTest* t = _initializer(params);
    t.group = self;
    return t;
}

+ (instancetype)groupForKey:(NSString*)key localizedDescription:(NSString*)description {
    RMBTTestQoSInitializerBlock initializer;

    if ([key isEqualToString:@"dns"]) {
        initializer = ^(NSDictionary *p) {
             return [[RMBTQoSDNSTest alloc] initWithParams:p];
        };
    } else if ([key isEqualToString:@"http_proxy"]) {
        initializer = ^(NSDictionary *p) {
            return [[RMBTQoSHTTPTest alloc] initWithParams:p];
        };
    } else if ([key isEqualToString:@"traceroute"]) {
        initializer = ^(NSDictionary *p) {
            return [[RMBTQoSTracerouteTest alloc] initWithParams:p masked:NO];
        };
    } else if ([key isEqualToString:@"traceroute_masked"]) {
        initializer = ^(NSDictionary *p) {
            return [[RMBTQoSTracerouteTest alloc] initWithParams:p masked:YES];
        };
    } else if ([key isEqualToString:@"website"]) {
        initializer = ^(NSDictionary *p) {
            return [[RMBTQoSWebTest alloc] initWithParams:p];
        };
    } else if ([key isEqualToString:@"udp"]) {
        initializer = ^(NSDictionary *p) {
            return [[RMBTQoSUDPTest alloc] initWithParams:p];
        };
    } else if ([key isEqualToString:@"tcp"]) {
        initializer = ^(NSDictionary *p) {
            return [[RMBTQoSTCPTest alloc] initWithParams:p];
        };
    } else if ([key isEqualToString:@"non_transparent_proxy"]) {
        initializer = ^(NSDictionary *p) {
            return [[RMBTQoSNonTransparentProxyTest alloc] initWithParams:p];
        };
    } else {
        [Log log:[NSString stringWithFormat:@"Unknown QoS group: %@", key]];
        return nil;
    }

    return [[RMBTQoSTestGroup alloc] initWithKey:key
                            localizedDescription:description
                                     initializer:initializer];
}

- (NSString*)description {
    return [NSString stringWithFormat:@"RMBTQoSTestGroup (key=%@)", _key];
}

@end

File: Sources/RMBTQoSTestGroup.swift
//
//  RMBTQoSTestGroup.swift
//  RMBT
//
//  Created by Sergey Glushchenko on 11.12.2021.
//  Copyright © 2021 appscape gmbh. All rights reserved.
//

import UIKit


@objc class RMBTQoSTestGroup: NSObject {

    @objc private(set) var key: String
    @objc private(set) var localizedDescription: String
    
    private var initializer: (_ p: [String: Any]) -> RMBTQoSTest?
    
    @objc(groupForKey:localizedDescription:)
    static func group(for key: String, description: String) -> Self? {
        let initializer: (_ p: [String: Any]) -> RMBTQoSTest?

        if key == "dns" {
            initializer = { p in
                return RMBTQoSDNSTest(with: p)
            }
        } else if key == "http_proxy" {
            initializer = { p in
                return RMBTQoSHTTPTest(with: p)
            }
        } else if key == "traceroute" {
            initializer = { p in
                return RMBTQoSTracerouteTest(with: p, masked: false)
            }
        } else if key == "traceroute_masked" {
            initializer = { p in
                return RMBTQoSTracerouteTest(with: p, masked: true)
            }
        } else if key == "website" {
            initializer = { p in
                return RMBTQoSWebTest(with: p)
            }
        } else if key == "udp" {
            initializer = { p in
                return RMBTQoSUDPTest(with: p)
            }
        } else if key == "tcp" {
            initializer = { p in
                return RMBTQoSTCPTest(with: p)
            }
        } else if key == "non_transparent_proxy" {
            initializer = { p in
                return RMBTQoSNonTransparentProxyTest(with: p)
            }
        } else {
            Log.logger.error("Unknown QoS group: \(key)")
            return nil
        }
        
        return RMBTQoSTestGroup(with: key, description: description, initializer: initializer) as? Self
    }
    
    @objc(testWithParams:)
    func test(with params: [String: Any]) -> RMBTQoSTest? {
        let t = initializer(params)
        t?.group = self
        return t
    }

    init(with key: String, description: String, initializer: @escaping (_ p: [String: Any]) -> RMBTQoSTest?) {
        self.key = key
        self.localizedDescription = description
        self.initializer = initializer
    }
    
    override var description: String {
        return String(format: "RMBTQoSTestGroup (key=\(String(describing: key)))")
    }
}
